PurchasesCoreSwift/Purchasing/PurchasesOrchestrator.swift

```diff
@@ -0,0 +1,508 @@
+                         payment: SKMutablePayment,
+                         presentedOfferingIdentifier maybePresentedOfferingIdentifier: String?,
+                         completion: @escaping PurchaseCompletedBlock) {
+        Logger.debug(String(format: "Make purchase called: %@", #function))
+        guard let productIdentifier = extractProductIdentifier(fromProduct: product, orPayment: payment) else {
+            Logger.error(Strings.purchase.could_not_purchase_product_id_not_found)
+            let errorMessage = "There was a problem purchasing the product: productIdentifier was nil"
+            completion(nil, nil, ErrorUtils.unknownError(message: errorMessage), false)
+            return
+        }
+
+        if !self.finishTransactions {
+            Logger.warn(Strings.purchase.purchasing_with_observer_mode_and_finish_transactions_false_warning)
+        }
+
+        payment.applicationUsername = appUserID
+        preventPurchasePopupCallFromTriggeringCacheRefresh(appUserID: appUserID)
+
+        if let presentedOfferingIdentifier = maybePresentedOfferingIdentifier {
+            Logger.purchase(String(format: Strings.purchase.purchasing_product_from_package,
+                                   productIdentifier,
+                                   presentedOfferingIdentifier))
+            lock.lock()
+            presentedOfferingIDsByProductID[productIdentifier] = presentedOfferingIdentifier
+            lock.unlock()
+
+        } else {
+            Logger.purchase(String(format: Strings.purchase.purchasing_product, productIdentifier))
+        }
+
+        productsManager.cacheProduct(product)
+
+        lock.lock()
+        defer {
+            lock.unlock()
+        }
+
+        guard purchaseCompleteCallbacksByProductID[productIdentifier] == nil else {
+            completion(nil, nil, ErrorUtils.operationAlreadyInProgressError(), false)
+            return
+        }
+        purchaseCompleteCallbacksByProductID[productIdentifier] = completion
+        storeKitWrapper.add(payment)
+    }
+
+}
+
+extension PurchasesOrchestrator: StoreKitWrapperDelegate {
+
+    public func storeKitWrapper(_ storeKitWrapper: StoreKitWrapper,
+                                updatedTransaction transaction: SKPaymentTransaction) {
+        switch transaction.transactionState {
+        case .restored, // for observer mode
+             .purchased:
+            handlePurchasedTransaction(transaction)
+        case .purchasing:
+            break
+        case .failed:
+            handleFailedTransaction(transaction)
+        case .deferred:
+            handleDeferredTransaction(transaction)
+        @unknown default:
+            Logger.warn("unhandled transaction state!")
+        }
+    }
+
+    public func storeKitWrapper(_ storeKitWrapper: StoreKitWrapper,
+                                removedTransaction transaction: SKPaymentTransaction) {
+        // todo: remove
+        // unused for now
+    }
```

Comment:
do we have an issue for this?

Reply:
We do now 😉  https://github.com/RevenueCat/purchases-ios/issues/751
Also dropped this in the code comment.